src/test/launchServer.ts

```diff
@@ -28,56 +28,38 @@ const DEFAULT_ENVIRONMENT_VARIABLES = {
 
 const newProcessLogPrefixer = () => new stream.Transform({
   transform(this: stream.Transform, chunk: Buffer, encoding: string, callback: stream.TransformCallback) {
-    this.push(chunk.toString().split(os.EOL).map((line: string): string => line ? `[WebServer] ${line}` : line).join(os.EOL));
+    this.push(chunk.toString().split(os.EOL).map((line: string): string => line ? `[Launch] ${line}` : line).join(os.EOL));
     callback();
   },
 });
 
-export class WebServer {
+class LaunchServer {
   private _killProcess?: () => Promise<void>;
   private _processExitedPromise!: Promise<any>;
-  constructor(private readonly config: WebServerConfig) { }
+  constructor(private readonly config: LaunchConfig) { }
 
-  public static async create(config: WebServerConfig): Promise<WebServer> {
-    const webServer = new WebServer(config);
-    if (config.port)
-      await webServer._verifyFreePort(config.port);
+  public static async create(config: LaunchConfig): Promise<LaunchServer> {
+    const launchServer = new LaunchServer(config);
     try {
-      const port = await webServer._startWebServer();
-      await webServer._waitForAvailability(port);
-      const baseURL = `http://localhost:${port}`;
-      process.env.PLAYWRIGHT_TEST_BASE_URL = baseURL;
-      console.log(`Using WebServer at '${baseURL}'.`);
-      return webServer;
+      await launchServer._startProcess();
+      await launchServer._waitForProcess();
+      return launchServer;
     } catch (error) {
-      await webServer.kill();
+      await launchServer.kill();
       throw error;
     }
   }
 
-  private async _verifyFreePort(port: number) {
-    const cancellationToken = { canceled: false };
-    const portIsUsed = await Promise.race([
-      new Promise(resolve => setTimeout(() => resolve(false), 100)),
-      waitForSocket(port, 100, cancellationToken),
-    ]);
-    cancellationToken.canceled = true;
-    if (portIsUsed)
-      throw new Error(`Port ${port} is used, make sure that nothing is running on the port`);
-  }
-
-  private async _startWebServer(): Promise<number> {
-    let collectPortResolve = (port: number) => { };
-    const collectPortPromise = new Promise<number>(resolve => collectPortResolve = resolve);
-    function collectPort(data: Buffer) {
-      const regExp = /http:\/\/localhost:(\d+)/.exec(data.toString());
-      if (regExp)
-        collectPortResolve(parseInt(regExp[1], 10));
-    }
-
+  private async _startProcess(): Promise<void> {
     let processExitedReject = (error: Error) => { };
     this._processExitedPromise = new Promise((_, reject) => processExitedReject = reject);
 
+    if (this.config.waitForPort) {
+      const portIsUsed = !await canBindPort(this.config.waitForPort);
+      if (portIsUsed && this.config.strict)
```

Comment:
I thought that when `portIsUsed`, but not `strict`, we'll just use the port without launching. I am probably missing the code for this. Do we have a test?

Reply:
Indeed, good catch! just pushed a new revision which fixed that a few seconds before your review.